File: packages/core/src/api/CatalogRegistry.ts
import { BlockCategory } from '../model/BlockCategory.js';
import type { BlockDefinition } from '../model/BlockDefinition.js';
import type {
  CamelComponentMeta,
  CamelComponentDescriptor,
  CamelCatalogProperty,
} from '../model/CamelCatalog.js';
import componentsCatalog from '../catalog/camel/components.json' with { type: 'json' };
import catalogVersion from '../catalog/camel/version.json' with { type: 'json' };
import blocks from '../catalog/blocks.json' with { type: 'json' };
import eips from '../catalog/eips.json' with { type: 'json' };

/** One entry from camel-catalog Maven JAR (Karavan components.json item). */
export interface CatalogComponentEntry {
  component: CamelComponentMeta;
  properties?: Record<string, RawCatalogProperty>;
  headers?: Record<string, unknown>;
}

interface RawCatalogProperty {
  displayName?: string;
  label?: string;
  type?: string;
  required?: boolean;
  secret?: boolean;
  defaultValue?: string | number | boolean;
  enum?: string[];
  description?: string;
  group?: string;
  kind?: string;
  deprecated?: boolean;
}

const blockList = blocks as BlockDefinition[];
const byType = new Map<string, BlockDefinition>(blockList.map((b) => [b.type, b]));

const entries = componentsCatalog as unknown as CatalogComponentEntry[];
const byScheme = new Map<string, CatalogComponentEntry>();
const byName = new Map<string, CatalogComponentEntry>();

for (const entry of entries) {
  const scheme = entry.component.scheme;
  const name = entry.component.name;
  byScheme.set(scheme, entry);
  byName.set(name, entry);
}

/** Camel catalog version from Maven (see packages/core/src/catalog/camel/version.json). */
export const CAMEL_CATALOG_VERSION = catalogVersion.version as string;
export const CAMEL_CATALOG_MAVEN = catalogVersion.maven as string;

function toCatalogProperties(
  props?: Record<string, RawCatalogProperty>
): CamelCatalogProperty[] {
  if (!props) return [];
  return Object.entries(props)
    .filter(([, v]) => !v.deprecated)
    .map(([name, v]) => ({
      name,
      displayName: v.displayName ?? name,
      type: v.type ?? 'string',
      required: Boolean(v.required),
      secret: Boolean(v.secret),
      defaultValue: v.defaultValue,
      enum: v.enum,
      description: v.description,
      label: typeof v.label === 'string' ? v.label : undefined,
      group: v.group,
      kind: v.kind,
    }));
}

function resolveEntry(schemeOrName: string): CatalogComponentEntry | undefined {
  return byScheme.get(schemeOrName) ?? byName.get(schemeOrName);
}

export function getCatalogComponent(scheme: string): CamelComponentDescriptor | undefined {
  const entry = resolveEntry(scheme);
  if (!entry) return undefined;
  return {
    scheme: entry.component.scheme,
    component: entry.component,
    properties: toCatalogProperties(entry.properties),
  };
}

export function getAllCatalogSchemes(): string[] {
  return Array.from(byScheme.keys()).sort();
}

export function getAllCatalogComponents(): CatalogComponentEntry[] {
  return entries;
}

export function rolesForScheme(scheme: string): BlockCategory[] {
  const c = resolveEntry(scheme)?.component;
  if (!c) return [];
  if (c.consumerOnly) return [BlockCategory.SOURCE];
  if (c.producerOnly) return [BlockCategory.DESTINATION];
  return [BlockCategory.SOURCE, BlockCategory.DESTINATION];
}

/** Spring Boot starter artifact (from catalog component.artifactId). */
export function getMavenStarter(blockType: string): string | null {
  const block = byType.get(blockType);
  if (!block) return null;
  if (block.category === BlockCategory.ACTION && !block.scheme) {
    if (getEipType(blockType) === 'json-xml') return 'camel-jacksonxml-starter';
    if (block.camelComponent === 'camel-core') return null;
  }
  const scheme = block.scheme;
  if (!scheme) return legacyStarter(block.camelComponent);
  const meta = resolveEntry(scheme)?.component;
  if (!meta) return legacyStarter(block.camelComponent);
  const id = meta.artifactId;
  if (!id || id === 'camel-core') return null;
  return id.endsWith('-starter') ? id : `${id}-starter`;
}

function legacyStarter(camelComponent: string): string | null {
  const map: Record<string, string | null> = {
    'camel-ftp-starter': 'camel-ftp-starter',
    'camel-kafka': 'camel-kafka-starter',
    'camel-activemq': 'camel-activemq-starter',
    'camel-undertow': 'camel-undertow-starter',
    'camel-file': 'camel-file-starter',
    'camel-mail': 'camel-mail-starter',
    'camel-jdbc': 'camel-jdbc-starter',
    'camel-http': 'camel-http-starter',
    'camel-xslt': 'camel-xslt-starter',
    'camel-jacksonxml': 'camel-jacksonxml-starter',
    'camel-timer': 'camel-timer-starter',
    'camel-core': null,
  };
  return map[camelComponent] ?? null;
}

export function enrichBlock(block: BlockDefinition): BlockDefinition & { catalog?: CamelComponentMeta } {
  if (!block.scheme) return block;
  const catalog = resolveEntry(block.scheme)?.component;
  return catalog ? { ...block, catalog } : block;
}

export function getBlockWithCatalog(type: string): (BlockDefinition & { catalog?: CamelComponentMeta }) | undefined {
  const block = byType.get(type);
  return block ? enrichBlock(block) : undefined;
}

export function getEipType(type: string): string | undefined {
  const row = (eips as { type: string; eipId: string }[]).find((e) => e.type === type);
  return row?.eipId;
}

File: java/flowcamel-core/src/main/java/com/flowcamel/core/registry/CatalogRegistry.java
package com.flowcamel.core.registry;

import com.flowcamel.core.catalog.CamelCatalogProperty;
import com.flowcamel.core.catalog.CamelComponentMeta;
import com.flowcamel.core.catalog.CatalogComponentEntry;
import com.flowcamel.core.catalog.CatalogJson;
import com.flowcamel.core.model.BlockCategory;
import com.flowcamel.core.model.BlockDefinition;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class CatalogRegistry {
  public static final String CAMEL_CATALOG_VERSION =
      CatalogJson.readTree("catalog/camel/version.json").get("version").asText();
  public static final String CAMEL_CATALOG_MAVEN =
      CatalogJson.readTree("catalog/camel/version.json").get("maven").asText();

  private static final List<CatalogComponentEntry> ENTRIES = CatalogJson.loadComponents();
  private static final Map<String, CatalogComponentEntry> BY_SCHEME = new HashMap<>();
  private static final Map<String, CatalogComponentEntry> BY_NAME = new HashMap<>();
  private static final Map<String, String> EIP_BY_TYPE = loadEips();

  static {
    for (CatalogComponentEntry e : ENTRIES) {
      BY_SCHEME.put(e.component.scheme, e);
      BY_NAME.put(e.component.name, e);
    }
  }

  private CatalogRegistry() {}

  private static Map<String, String> loadEips() {
    JsonNode arr = CatalogJson.readTree("catalog/eips.json");
    Map<String, String> map = new HashMap<>();
    for (JsonNode row : arr) map.put(row.get("type").asText(), row.get("eipId").asText());
    return map;
  }

  public static Optional<CatalogComponentEntry> resolveEntry(String schemeOrName) {
    return Optional.ofNullable(BY_SCHEME.get(schemeOrName)).or(() -> Optional.ofNullable(BY_NAME.get(schemeOrName)));
  }

  public static List<CamelCatalogProperty> getCatalogProperties(String scheme) {
    return resolveEntry(scheme)
        .map(e -> toCatalogProperties(e.properties))
        .orElse(List.of());
  }

  private static List<CamelCatalogProperty> toCatalogProperties(Map<String, JsonNode> props) {
    if (props == null) return List.of();
    List<CamelCatalogProperty> out = new ArrayList<>();
    for (var e : props.entrySet()) {
      JsonNode v = e.getValue();
      if (v.has("deprecated") && v.get("deprecated").asBoolean()) continue;
      CamelCatalogProperty p = new CamelCatalogProperty();
      p.name = e.getKey();
      p.displayName = text(v, "displayName", p.name);
      p.type = text(v, "type", "string");
      p.required = v.has("required") && v.get("required").asBoolean();
      p.secret = v.has("secret") && v.get("secret").asBoolean();
      if (v.has("defaultValue") && !v.get("defaultValue").isNull()) {
        JsonNode dv = v.get("defaultValue");
        p.defaultValue = dv.isBoolean() ? dv.asBoolean() : dv.asText();
      }
      if (v.has("enum")) {
        List<String> en = new ArrayList<>();
        for (JsonNode item : v.get("enum")) en.add(item.asText());
        p.enumValues = en.toArray(String[]::new);
      }
      p.description = text(v, "description", null);
      p.label = v.has("label") && v.get("label").isTextual() ? v.get("label").asText() : null;
      p.kind = text(v, "kind", null);
      out.add(p);
    }
    return out;
  }

  private static String text(JsonNode n, String field, String def) {
    return n.has(field) && !n.get(field).isNull() ? n.get(field).asText() : def;
  }

  public static List<String> getAllCatalogSchemes() {
    return BY_SCHEME.keySet().stream().sorted().toList();
  }

  public static Optional<String> getEipType(String blockType) {
    return Optional.ofNullable(EIP_BY_TYPE.get(blockType));
  }

  public static String getMavenStarter(String blockType) {
    BlockDefinition block = BlockRegistry.getBlock(blockType).orElse(null);
    if (block == null) return null;
    if (block.category == BlockCategory.ACTION && block.scheme == null) {
      if ("json-xml".equals(getEipType(blockType).orElse(null))) return "camel-jacksonxml-starter";
      if ("camel-core".equals(block.camelComponent)) return null;
    }
    if (block.scheme == null) return legacyStarter(block.camelComponent);
    CatalogComponentEntry entry = resolveEntry(block.scheme).orElse(null);
    if (entry == null) return legacyStarter(block.camelComponent);
    String id = entry.component.artifactId;
    if (id == null || id.isEmpty() || "camel-core".equals(id)) return null;
    return id.endsWith("-starter") ? id : id + "-starter";
  }

  private static String legacyStarter(String camelComponent) {
    return switch (camelComponent) {
      case "camel-ftp-starter" -> "camel-ftp-starter";
      case "camel-kafka" -> "camel-kafka-starter";
      case "camel-activemq" -> "camel-activemq-starter";
      case "camel-undertow" -> "camel-undertow-starter";
      case "camel-file" -> "camel-file-starter";
      case "camel-mail" -> "camel-mail-starter";
      case "camel-jdbc" -> "camel-jdbc-starter";
      case "camel-http" -> "camel-http-starter";
      case "camel-xslt" -> "camel-xslt-starter";
      case "camel-jacksonxml" -> "camel-jacksonxml-starter";
      case "camel-timer" -> "camel-timer-starter";
      default -> null;
    };
  }
}
